### ISSUE TITLE
Qwen3-Coder template: model fails to trigger XML tool calls with large prompts — request JSON format option

### ISSUE THREAD

**Issue by rykerzhou** [ggml-org/llama.cpp#26530]
[llama_cpp_issue.md](https://github.com/user-attachments/files/30672460/llama_cpp_issue.md)
**Title: Qwen3-Coder template: model often fails to trigger XML tool calls with large prompts — request JSON format option**

## Summary

Qwen3-Coder models (Qwen3.5/Qwen3.6 based) using the built-in specialized template often **fail to output `<tool_call>`** when presented with large system prompts and many tools (e.g., 20+ tools, 10K+ token prompts typically sent by AI agent frameworks). The model responds with plain text instead of making tool calls, even though the lazy PEG-native grammar is correctly generated and waiting.

The issue became visible after PR #24869 (AC parser, merged Jun 21 2026) improved XML grammar generation. Older builds (b8646, Apr 3 2026) accidentally worked around this because the old delimiter parser would fail on complex multi-tool grammars and fall back to a JSON array grammar — the JSON tool instructions were more reliably followed by the model.

## Steps to Reproduce

1. Start llama-server with a Qwen3.5-based GGUF model:
   ```
   llama-server -m Qwen3.5-model.gguf --reasoning off --jinja --port 1234
   ```
2. Send a `/v1/chat/completions` request with 20+ tool definitions and a long system prompt (10K+ tokens total), `tool_choice: "auto"` (default)
3. The model responds with plain text content (e.g., "I'll help you create...") with `finish_reason: "stop"` instead of `finish_reason: "tool_calls"`

The lazy grammar is correctly generated but **never triggered** — `Grammar still awaiting trigger after token ... (<|im_end|>)`.

For comparison, the identical request with `tool_choice: "required"` works correctly.

## Root Cause

The Qwen3-Coder template injects XML-format tool instructions (`<function=name>`, `<parameter=x>`) into the system prompt. The model, when faced with large/complex prompts, tends to output text instead of voluntarily initiating the XML tool call format.

The old parser (pre-#24869) had a bug where complex XML grammars with many tools would fail to parse:
```
parse: error parsing grammar: number of rules that are going to be repeated multiplied by the new repetition exceeds sa...
```
This failure caused a fallback to JSON array grammar format, whose corresponding tool instructions happened to be better followed by the model.

## Proposed Solution

Add support for a `tool_call_format` option in the Qwen3-Coder specialized template (controllable via `--chat-template-kwargs`), allowing users to choose `"json"` format for tool calls instead of the default XML. For example:

```
--chat-template-kwargs '{"tool_call_format": "json"}'
```

This would:
1. Generate JSON-format tool instructions in the system prompt
2. Use the JSON array grammar (like the old accidental fallback)
3. Improve tool call reliability for Qwen3-Coder models in agent scenarios

## Environment

- llama.cpp build b10237 (2b63e0610)
- Model: Qwen3.5-9B based (CoPaw-Flash-9B-Q4_K_M, `qwen35` architecture)
- ROCm backend, AMD Radeon 8060S
- Windows 11

**Comment by rykerzhou** [comment]
## Update: Vulkan backend works on the same build

I tested the **exact same build b10237** with the **Vulkan backend** (same model, same parameters), and tool calls work correctly. This means the issue may be a **ROCm-specific regression** rather than a pure template problem.

| Build | Backend | Result |
|-------|---------|--------|
| b8646 | ROCm | ✅ Tool calls work |
| b10237 | ROCm | ❌ Model outputs text, never triggers `<tool_call>` |
| b10237 | Vulkan | ✅ Tool calls work |

I'm puzzled why the GPU backend would affect tool call triggering. The template rendering, grammar generation, and generation prompt are identical across backends. But somehow the model's token-by-token output differs — on ROCm it chooses text tokens, on Vulkan it chooses `<tool_call>`.

Possible ROCm-specific causes between b8646 → b10237:
1. Numerical precision changes affecting token logits (even tiny shifts can change the first token choice)
2. KV cache / memory handling changes with large prompts (`--no-mmap` is now deprecated)
3. ROCm graph optimization producing different computation results

Verbose logs confirm the grammar is correctly set up and waiting on both backends — the difference is purely in whether the model decides to output `<tool_call>`. This rules out template/grammar as the sole cause.

Would appreciate guidance on how to bisect this further, or any known ROCm changes between these builds that could affect token probabilities.
